---
title: "Aggregte studies"
author: "Kent Riemondy RBI"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
```

```{r}
library(ComplexHeatmap)
library(tidyverse)
library(Seurat)
library(here)
library(readxl)
library(scbp) # devtools::install_github("rnabioco/scbp")
library(openxlsx)
library(cowplot)
data_dir <- here("data")
formatted_data_dir <- "processed_data_v2"
fig_dir <- "figs"

dir.create(fig_dir, showWarnings = FALSE)
dir.create(formatted_data_dir, showWarnings = FALSE)

obj_dir <- "objects"
dir.create(obj_dir)

select <- function(...){
  dplyr::select(...)
}

slice <- function(...){
  dplyr::slice(...)
}
```

## Get ortholog table

```{r}
ortholog_table <- file.path(data_dir, "orthologs", "mouse_human_orthologs.tsv")

if(!file.exists(ortholog_table)){
  dir.create(file.path(data_dir, "orthologs"), showWarnings = FALSE)
  
  library(biomaRt)
  
  ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  
  orthologs <- getBM(mart = ensembl, 
      attributes = c("external_gene_name",
                     "mmusculus_homolog_associated_gene_name"))
  
  orthologs <- filter(orthologs,
                      !is.na(mmusculus_homolog_associated_gene_name),
                      mmusculus_homolog_associated_gene_name != "") %>%
    group_by(external_gene_name) %>%
    mutate(n_orthos_h = n()) %>%
    group_by(mmusculus_homolog_associated_gene_name) %>%
    mutate(n_orthos_m = n()) %>%
    filter(n_orthos_h == 1, n_orthos_m == 1) %>%
    dplyr::select(-starts_with("n_orthos")) %>%
    ungroup()
  
  orthologs <- mutate(orthologs, 
                      ortho_id = str_c(external_gene_name, ":", mmusculus_homolog_associated_gene_name)) %>% 
    dplyr::rename(human_gene_name = external_gene_name,
                  mouse_gene_name = mmusculus_homolog_associated_gene_name)

  write_tsv(orthologs, ortholog_table)
}

orthologs <- read_tsv(
  ortholog_table,
  col_types = cols(
    human_gene_name = col_character(),
    mouse_gene_name = col_character(),
    ortho_id = col_character()
  )
)

```

## Gene lists

```{r}
studies <- c(
  "Adams_et_al_bioRxiv" = "Data_S1.xlsx",
  "Habermann_et_al_bioRxiv" = "KRT17_vs_allepi.csv",
  "Reyfman_et_al_ajrccm" = "3.tsv.gz",
  "Joshi_et_al_european-r-j" = "AT2_clean.markers.csv",
  "Kobayashi_et_al_ncb" = "41556_2020_542_MOESM2_ESM.xlsx",
  "Riemondy_et_al_JCI-insight" = "2.14.2019_Suppl_Table_5.xlsx",
  "Strunz_et_al_nc" = "41467_2020_17358_MOESM4_ESM.xlsx",
  "Wu_et_al_cell" = "Nan Tang Table S2.xlsx"
  )

studies <- imap(studies, ~file.path(data_dir, .y, .x))

combined_markers <- list()
combined_tables <- list()
combined_expression <- list()

mouse_studies <- names(studies)[4:8]
human_studies <- names(studies)[1:3]
```

## Adams_et_al_bioRxiv

```{r}
study_id <- "Adams_et_al_bioRxiv"
mkrs <- studies[[study_id]]

tbl <- read_excel(mkrs) %>% 
  filter(cellType == "Aberrant Basaloid") %>% 
  arrange(wilcox.fdr) %>% 
  filter(wilcox.fdr < 0.05,
         log1p_FC > 0)

mouse_markers <- left_join(tbl, orthologs, by = c("gene" = "human_gene_name")) %>% 
  pull(mouse_gene_name) %>% 
  .[!is.na(.)]

combined_markers[[study_id]] <- mouse_markers
combined_tables[[study_id]] <- tbl
```

## Habermann_et_al_bioRxiv

```{r}
study_id <- "Habermann_et_al_bioRxiv"
mkrs <- studies[[study_id]]

tbl <- read_csv(mkrs) %>% 
  arrange(p_val_adj) %>% 
  filter(p_val_adj < 0.05,
         avg_logFC > 0)

mouse_markers <- left_join(tbl, orthologs, by = c("X1" = "human_gene_name")) %>% 
  pull(mouse_gene_name) %>% 
  .[!is.na(.)]

combined_markers[[study_id]] <- mouse_markers
combined_tables[[study_id]] <- tbl

so <- readRDS(file.path(data_dir, study_id, "GSE135893_ILD_annotated_diet.rds"))
so_sub <- subset(so, celltype %in% c("KRT5-/KRT17+", "AT1", "AT2", "Transitional AT2")) 
so_sub <- NormalizeData(so_sub, assay = "RNA")
avg_expr <- AverageExpression(so_sub)$RNA
colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "human_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "human_gene_name")

combined_expression[[study_id]] <- avg_expr

rm(so, so_sub); gc()
```

## Reyfman_et_al_ajrccm

```{r}
study_id <- "Reyfman_et_al_ajrccm"
mkrs <- studies[[study_id]]

tbl <- read_tsv(mkrs) %>% 
  arrange(`p_val|float`, desc(`avg_diff|float`)) %>% 
  filter(`p_val|float` < 0.05,
         `avg_diff|float` > 0)

mouse_markers <- left_join(tbl, orthologs, by = c("symbol" = "human_gene_name")) %>% 
  pull(mouse_gene_name) %>% 
  .[!is.na(.)]

combined_markers[[study_id]] <- mouse_markers
combined_tables[[study_id]] <- tbl

so_fn <- file.path(data_dir, "Reyfman_et_al_ajrccm", "so.rds")

if(!file.exists(so_fn)){
  mat <- read_tsv(file.path(data_dir, "Reyfman_et_al_ajrccm", "exprMatrix.tsv.gz"))
  mat <- column_to_rownames(mat, "gene")
  mat <- as.data.frame(mat) %>% as.matrix()
  mat <- as(mat, "sparseMatrix")
  
  # mat_url <- "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3489182&format=file&file=GSM3489182%5FDonor%5F01%5Fraw%5Fgene%5Fbc%5Fmatrices%5Fh5%2Eh5"
  # mat_fn <- file.path(data_dir, "Reyfman_et_al_ajrccm", "matrix.h5")
  # download.file(mat_url, mat_fn)
  # so <- Read10X_h5(mat_fn)
  
  mdata <- read_tsv(file.path(data_dir, "Reyfman_et_al_ajrccm", "meta.tsv")) %>% 
    column_to_rownames("Cell")
  so <- CreateSeuratObject(mat, meta.data = mdata)
  saveRDS(so, so_fn)
}
so <- readRDS(so_fn)
Idents(so) <- "Cluster"
avg_expr <- AverageExpression(so)$RNA
colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "human_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "human_gene_name")

combined_expression[[study_id]] <- avg_expr
rm(so); gc()
```

## Joshi_et_al_european-r-j
 
```{r}
study_id <- "Joshi_et_al_european-r-j"
mkrs <- studies[[study_id]]

tbl <- read_csv(mkrs) %>% 
  filter(cluster == "AT2_3") %>% 
  arrange(p_val_adj) %>% 
  filter(p_val_adj < 0.05,
         avg_logFC > 0)

combined_markers[[study_id]] <-tbl$X1
combined_tables[[study_id]] <- tbl

so_fn <- file.path(data_dir, study_id, "so.rds")
if(!file.exists(so_fn)){
  load(file.path(data_dir, study_id, "AT2_clean.Robj"))
  AT2_clean$pub_clusters <- Idents(AT2_clean)
  
  # get data from UCSC cellbrowser
  mat <- read_tsv(file.path(data_dir, 
                            study_id, 
                            "data", 
                            "full-exprMatrix.tsv.gz"))
  mat <- column_to_rownames(mat, "gene") 
  mat <- as.matrix(mat)
  mat <- as(mat, "sparseMatrix")
  
  mdata <- read_tsv(file.path(data_dir, 
                            study_id, 
                            "data", 
                            "full-meta.tsv"))
  mdata <- column_to_rownames(mdata, "Cell") %>% 
    as.data.frame()
  
  so <- CreateSeuratObject(mat, meta.data = mdata)
  at1 <- subset(so, subset = Cluster == "AT1")
  at1$pub_clusters <- "AT1"
  at1@meta.data <- at1@meta.data[, c("orig.ident", "pub_clusters")]  
  so <- merge(AT2_clean, at1)
  Idents(so) <- "pub_clusters"
  saveRDS(so, so_fn)
}

so <- readRDS(so_fn)
avg_expr <-  AverageExpression(so)$RNA
#avg_expr <- avg_expr[, "AT2_3", drop = FALSE]
colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "mouse_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "mouse_gene_name")

combined_expression[[study_id]] <- avg_expr

rm(so); gc()
```


## Kobayashi_et_al_ncb

```{r}
study_id <- "Kobayashi_et_al_ncb"
mkrs <- studies[[study_id]]

tbl <- read_excel(mkrs, skip = 1) %>% 
  .[, 1:4]
tbl <- list(tbl[[1]], tbl[[2]], tbl[[3]], tbl[[4]])  %>% 
  map(~.x[!is.na(.x)]) 
names(tbl) <- c("AEC2", "AEC2-proliferating", "PATS", "AEC1")
tbl <- map_dfr(tbl, ~tibble(gene = .x), .id = "cell_type")


combined_markers[[study_id]] <- tbl %>%
  filter(cell_type == "PATS") %>% 
  pull(gene)

combined_tables[[study_id]] <- tbl


so_fn <- file.path(data_dir, study_id, "so.rds")
if(!file.exists(so_fn)){
  mat_fn <- file.path(data_dir, study_id, "GSM4210295_MTECplus.tsv.gz")
  if(!file.exists(mat_fn)) {
    download.file("https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4210nnn/GSM4210295/suppl/GSM4210295_MTECplus.tsv.gz", mat_fn)
  }
  mat <- read_tsv(mat_fn) %>% 
    as.data.frame() %>% 
    column_to_rownames("GENE") %>% 
    as.matrix() %>% 
    as("sparseMatrix")
  mdata <- readxl::read_excel(file.path(data_dir,
                                        study_id,
                                        "epithelial cells_MTEC_organoids_d10_active_ident.xlsx")) %>% 
    mutate(cell_id = str_split(Cell_barcode, ":", simplify = TRUE) %>% 
             .[, 2] %>% 
             str_remove("x$")) %>% 
    column_to_rownames("cell_id") %>% 
    select(-Cell_barcode,
           cell_type = active.ident) %>% 
    as.data.frame(stringsAsFactors = FALSE)
  so <- CreateSeuratObject(mat, meta.data = mdata)
  so$cell_type <- ifelse(is.na(so$cell_type), "other", so$cell_type)
  
  so <- NormalizeData(so)
  
  saveRDS(so, so_fn)
}

so <- readRDS(so_fn)
so <- subset(so, subset = cell_type != "other")
Idents(so) <- "cell_type"
avg_expr <- AverageExpression(so)$RNA


colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "mouse_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "mouse_gene_name")
combined_expression[[study_id]] <- avg_expr
```

## Riemondy_et_al_JCI-insight

```{r}
study_id <- "Riemondy_et_al_JCI-insight"
mkrs <- studies[[study_id]]

tbl <- read_excel(mkrs, sheet = 2,
                  col_names = c(
                    "Gene",
                    "avg_logFC",
                    "pct1",
                    "pct2",
                    "pval",
                    "adj_p_val"
                  ), skip = 1) %>% 
  arrange(adj_p_val) %>% 
  filter(adj_p_val < 0.05)

combined_markers[[study_id]] <-tbl$Gene
combined_tables[[study_id]] <- tbl

so_fn <- file.path(data_dir, study_id, "so.rds")
if(!file.exists(so_fn)){
  ## update old seurat file
  so <- readRDS(file.path(data_dir, study_id, "s_obj_final.rds"))
  so <- UpdateSeuratObject(so)
  mdata <- so@meta.data
  new_mdata <- tibble::rownames_to_column(mdata, "cell") %>% 
    dplyr::mutate(sample_type = ifelse(str_detect(new_expt_id, 
                                                  "Non-ATII epithelial"),
                                       "Naive Non-AEC2 Epithelial",
                                       ifelse(str_detect(new_expt_id,
                                                         "ATII expt"),
                                              "Naive AEC2",
                                              "Injured AEC2")),
                  experiment = ifelse(expt == "expt1", 
                                      "Experiment 1",
                                      "Experiment 2")) %>% 
    dplyr::select(cell, sample_type, experiment) %>% 
    tibble::column_to_rownames("cell")
  
  so <- AddMetaData(so, new_mdata)
  
  so <- AddMetaData(so, 
                    data.frame(row.names = rownames(so@meta.data),
                               technical_replicate = str_match(so@meta.data$sample_names, "([0-9])expt")[, 2]))
  
  so <- AddMetaData(so,
                    data.frame(row.names = rownames(so@meta.data),
                               experiment_and_tech_rep = str_c(so@meta.data$experiment,
                                                               ": Technical Replicate ",
                                                               so@meta.data$technical_replicate)))
  
  cell_id_relabel <- c("Endothelial" = "Endothelial/Fibroblast",
                       "Basal" = "Basal",
                       "Club" = "Club",
                       "Ciliary" = "Ciliated",
                       "Naive Type I" = "Naive AEC1",
                       "Macrophages" = "Macrophage",
                       "Injured Type II" = "Other Injured AEC2",
                       "Proliferating Type II" = "Injured AEC2: Proliferating",
                       "Cell Cycle Arrest Type II" = "Injured AEC2: Early Transdifferentiating",
                       "Transdifferentiating Type II" = "Injured AEC2: Late Transdifferentiating",
                       "Naive Type II" = "Naive AEC2")
  
  
  m_df <- data.frame(row.names = rownames(so@meta.data),
                     pretty_cell_labels = cell_id_relabel[so@meta.data$labeled_clusters])
  
  m_df$pretty_cell_labels <- factor(m_df$pretty_cell_labels,
                                    levels = cell_id_relabel)
  
  so <- AddMetaData(so, m_df)
  saveRDS(so, file.path(data_dir, study_id, "so.rds"))
}

so <- readRDS(so_fn)
Idents(so) <- "pretty_cell_labels"
avg_expr <-  AverageExpression(so)$RNA
avg_expr <- avg_expr[, c("Naive AEC1",
                         "Naive AEC2",
                         "Injured AEC2: Early Transdifferentiating",
                         "Injured AEC2: Late Transdifferentiating", 
                         "Injured AEC2: Proliferating", 
                         "Other Injured AEC2"), drop = FALSE]
colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "mouse_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "mouse_gene_name")

combined_expression[[study_id]] <- avg_expr

rm(so); gc()
```

## Strunz_et_al_nc

```{r}
study_id <- "Strunz_et_al_nc"
mkrs <- studies[[study_id]]

tbl <- read_excel(mkrs, sheet = 2) %>% 
  filter(`annotated cluster` ==	"Krt8+ ADI") %>% 
  arrange(p_val_adj) %>% 
  filter(p_val_adj < 0.05,
         avg_logFC > 0)

combined_markers[[study_id]] <- tbl$gene
combined_tables[[study_id]] <- tbl


so_fn <- file.path(data_dir, study_id, "so_whole.rds")
if(!file.exists(so_fn)){
  
  mtx <- Matrix::readMM(gzcon(url("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nnn/GSE141259/suppl/GSE141259_WholeLung_rawcounts.mtx.gz")))
  rownames(mtx) <- readLines(gzcon(url("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nnn/GSE141259/suppl/GSE141259_WholeLung_genes.txt.gz")))
  colnames(mtx) <- readLines(gzcon(url("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nnn/GSE141259/suppl/GSE141259_WholeLung_barcodes.txt.gz")))
  
  mdata <- read_csv("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nnn/GSE141259/suppl/GSE141259_WholeLung_cellinfo.csv.gz") %>% 
    select(X1, id.orig.ident = orig.ident, identifier, res.2:spline_cluster) %>% 
    column_to_rownames("X1") %>% 
    as.data.frame()  
  
  so_whole_lung <- CreateSeuratObject(mtx, meta.data = mdata) %>% 
    NormalizeData()
  
  
  mtx <- Matrix::readMM(gzcon(url("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nnn/GSE141259/suppl/GSE141259_HighResolution_rawcounts.mtx.gz")))
  rownames(mtx) <- readLines(gzcon(url("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nnn/GSE141259/suppl/GSE141259_HighResolution_genes.txt.gz")))
  colnames(mtx) <- readLines(gzcon(url("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nnn/GSE141259/suppl/GSE141259_HighResolution_barcodes.txt.gz")))
  
  mdata <- read_csv("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nnn/GSE141259/suppl/GSE141259_HighResolution_cellinfo.csv.gz") %>% 
    select(X1, sample, grouping) %>% 
    column_to_rownames("X1") %>% 
    as.data.frame()  
  
  so_high_res <- CreateSeuratObject(mtx, meta.data = mdata) %>% 
    NormalizeData()
  # labels are missing for the hiRes data, only sample labels
  # took scanpy oject ("hiRes.hdf5) and wrote meta data output
  mdata <- read_csv(file.path(data_dir, study_id, "scanpy_metadata.csv")) %>% 
    mutate(cell_id = str_c(sample_batch, index)) %>% 
    select(cell_id, name, grouping, index, sample_batch, cell_type, cell_type_res4, cell_type_recombined) %>% 
    column_to_rownames("cell_id") %>% 
    .[intersect(rownames(.), colnames(so_high_res)), ]
  
  so_high_res <- AddMetaData(so_high_res, mdata)
  saveRDS(so_high_res,  file.path(data_dir, study_id, "so_high_res.rds"))
  saveRDS(so_whole_lung,  file.path(data_dir, study_id, "so_whole.rds"))
  
  
}
so_hires <- readRDS(file.path(data_dir, study_id, "so_high_res.rds"))
Idents(so_hires) <- "cell_type_recombined"
avg_expr_hires <-  AverageExpression(so_hires)$RNA[,
                                       c("Krt8 ADI",
                                         "AT1 cells",
                                         "AT2 cells",
                                         "activated AT2","proliferation")]

colnames(avg_expr_hires) <- paste0("time_course:", colnames(avg_expr_hires))

so_whole <- readRDS(file.path(data_dir, study_id, "so_whole.rds"))
Idents(so_whole) <- "cell.type"
avg_expr_whole <-  AverageExpression(so_whole)$RNA[,
                                       c("Krt8 ADI", "Activated AT2 cells", "AT1 cells", "AT2 cells")]

colnames(avg_expr_whole) <- paste0("whole_lung:", colnames(avg_expr_whole))

shared_genes <- intersect(rownames(avg_expr_whole), rownames(avg_expr_hires))
avg_expr <- cbind(avg_expr_whole[shared_genes, ], avg_expr_hires[shared_genes, ])

colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "mouse_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "mouse_gene_name")

combined_expression[[study_id]] <- avg_expr

so_hires$cell_type <- so_hires$cell_type_recombined
so_hires$dataset <- "time-course"
so_whole$cell_type <- so_whole$cell.type
so_whole$dataset <- "whole-lung"
so <- merge(so_hires, so_whole)

so$cell_type <- ifelse(so$cell_type == "Activated AT2 cells",
                       "activated AT2",
                       so$cell_type)

so$pretty_cell_labels <- str_c("Bleo:", so$dataset, "-", so$cell_type)

saveRDS(so, file.path(data_dir, study_id, "so.rds"))

rm(so_hires, so_whole)
```


## Wu_et_al_cell

```{r}
study_id <- "Wu_et_al_cell"
mkrs <- studies[[study_id]]

tbl <- read_excel(mkrs, skip = 1) %>% 
  filter(`fold change (subpopulation I AT2 cells/subpopulation II AT2 cells)` > 0)

combined_markers[[study_id]] <- tbl$`Gene name`
combined_tables[[study_id]] <- tbl

avg_expr_d7 <- read_excel(file.path(data_dir, study_id, "C7.average.exp.xlsx"))
avg_expr_d7 <- dplyr::rename(avg_expr_d7, 
                          mouse_gene_name = "...1")

colnames(avg_expr_d7)[2:3] <- paste0(study_id, "::", colnames(avg_expr_d7)[2:3], "_d7")

so_fn <- file.path(data_dir, study_id, "so.rds")
if(!file.exists(so_fn)){
  # also process d21 data
  mat <- read_csv(file.path(data_dir,  study_id, "matrix.tsv.gz"))
  mat <- column_to_rownames(mat, "X1") 
  mat <- as.matrix(mat)
  mat <- as(mat, "sparseMatrix")
  
  mdata <- read_csv(file.path(data_dir, study_id, "meta_info_C0_C7_C21.csv")) %>% 
    dplyr::select(cell = X1, sample_name, cell_type) %>% 
    filter(cell %in% colnames(mat)) %>% 
    as.data.frame() %>% 
    column_to_rownames("cell") 
  
  mat <- mat[, rownames(mdata)]
  so <- CreateSeuratObject(mat, meta.data = mdata) %>% 
    NormalizeData()
  
    # also process at1 cells from previous pub (no metadata :( )
  at1_mat <- read_csv(file.path(data_dir,  study_id, "at1_data", "GSM2858341_AT1_P60.exprs.csv"))
  at1_mat <- column_to_rownames(at1_mat, "X1") %>% 
    as.matrix(.) %>% 
    as(., "sparseMatrix")
  so_at1 <- CreateSeuratObject(at1_mat)
  so_at1 <- FindVariableFeatures(so_at1) %>% 
    ScaleData() %>% 
    RunPCA(seed.value = 42) %>% 
    RunUMAP(dims = 1:10, seed.value = 42) %>% 
    FindNeighbors() %>% 
    FindClusters(resolution = c(0.05, 0.1, 0.2), random.seed = 42)
  so_at1$cell_type <- ifelse(so_at1$RNA_snn_res.0.2 %in% c("0", "1", "2", "4"),
                             "AT1",
                             "other")
  so_at1 <- subset(so_at1, 
                   subset = cell_type == "AT1")
  # get rid of a few outliers
  to_drop <- c("CAGGTGCCATTACGAC",
    "CTAGCCTGTGTTTGGT",
    "GTACGTAGTTCCGGCA",
    "TGGGAAGAGAAGAAGC")

  so_at1 <- subset(so_at1, cells = to_drop, invert = TRUE)
  
  so <- merge(so, so_at1)
  
  saveRDS(so, so_fn)
}

so <- readRDS(so_fn)
Idents(so) <- "cell_type"
avg_expr <- AverageExpression(so)$RNA
#avg_expr <- avg_expr[, "population II", drop = FALSE]
colnames(avg_expr)[1:2] <- paste0(study_id, "::", colnames(avg_expr)[1:2], "_d21") %>% 
  str_replace_all(" ", "_")
colnames(avg_expr)[3] <- paste0(study_id, "::", colnames(avg_expr[3]))
avg_expr <- rownames_to_column(avg_expr, "mouse_gene_name")
avg_expr_d21 <- left_join(avg_expr, orthologs, by = "mouse_gene_name")
avg_expr <- left_join(avg_expr_d21, avg_expr_d7, by = "mouse_gene_name")
combined_expression[[study_id]] <- avg_expr
```


## Save 

```{r}

saveRDS(combined_markers, file.path(formatted_data_dir, "gene_lists.rds"))
saveRDS(combined_tables, file.path(formatted_data_dir, "gene_tables.rds"))
saveRDS(combined_expression, file.path(formatted_data_dir, "avg_expression.rds"))

```

```{r}

combined_markers <- readRDS(file.path('processed_data', "gene_lists.rds"))

library(eulerr)
#combined_markers <- map(combined_markers, ~.x[1:98])
fit <- venn(combined_markers[mouse_studies])
pdf(file.path(fig_dir, "marker_overlap_mouse_studies.pdf"))
plot(fit, 
     labels = TRUE,
     quantities = TRUE,
     fills = list(fill = palette_OkabeIto[1:4],
                  alpha = 0.5))
dev.off()


fit <- venn(map(combined_markers[mouse_studies], ~.x[1:50]))
pdf(file.path(fig_dir, "marker_overlap_mouse_studies_pvalue_top50.pdf"))
plot(fit, 
     labels = TRUE,
     quantities = TRUE,
     fills = list(fill = palette_OkabeIto[1:4],
                  alpha = 0.5))
dev.off()


fit <- euler(combined_markers[human_studies])
pdf(file.path(fig_dir, "marker_overlap_human_studies.pdf"))
plot(fit, 
     labels = TRUE,
     quantities = TRUE,
     fills = list(fill = palette_OkabeIto[1:6],
                  alpha = 0.5))
dev.off()


markers_shared <- Reduce(intersect, combined_markers[mouse_studies])

all_markers <- unlist(combined_markers[mouse_studies], use.names = FALSE) %>% unique()

                      
mrkr_lgl <- imap(combined_markers[mouse_studies],
    function(x, id) {
      df <- data.frame(tmp = all_markers %in% x, 
                       row.names = all_markers)
      colnames(df) <- id
      df}) %>% 
  do.call(cbind, .)

markers_3_of_4 <- mrkr_lgl[rowSums(mrkr_lgl) == 3, ] %>% 
  rownames_to_column("gene")

openxlsx::write.xlsx(list(shared_markers = markers_shared,
     shared_3_of_4 = markers_3_of_4),
     file.path(fig_dir, "shared_genes.xlsx"))
```

```{r, eval = FALSE}
combined_tables <- readRDS(file.path('processed_data', "gene_tables.rds"))
mouse_markers_fc <- list()

mouse_markers_fc$`Joshi_et_al_european-r-j`  <- combined_tables$`Joshi_et_al_european-r-j` %>% 
  filter(p_val_adj < 0.05) %>% 
  arrange(desc(avg_logFC)) %>% 
  pull(1)

mouse_markers_fc$`Riemondy_et_al_JCI-insight` <- combined_tables$`Riemondy_et_al_JCI-insight` %>% 
  filter(adj_p_val < 0.05) %>% 
  arrange(desc(avg_logFC)) %>% 
  pull(1)

mouse_markers_fc$Strunz_et_al_bioRxiv <- combined_tables$Strunz_et_al_bioRxiv %>% 
    filter(p_val_adj < 0.05) %>% 
  arrange(desc(avg_logFC)) %>% 
  pull(1)

mouse_markers_fc$Wu_et_al_cell <- combined_tables$Wu_et_al_cell %>% 
  arrange(desc(`fold change (subpopulation I AT2 cells/subpopulation II AT2 cells)`)) %>% 
  pull(1)



fit <- venn(map(mouse_markers_fc, ~.x[1:50]))
pdf(file.path(fig_dir, "marker_overlap_mouse_studies_fold_change_top50.pdf"))
plot(fit, 
     labels = TRUE,
     quantities = TRUE,
     fills = list(fill = palette_OkabeIto[1:4],
                  alpha = 0.5))
dev.off()

```


```{r, eval = FALSE}
library(gprofiler2)

if(!file.exists("go_results.rds")) {
  to_profile <- c(
    combined_markers[mouse_studies], 
    list(conserved_markers = markers_shared),
    list(mostly_conserved_markers = markers_3_of_4$gene))
  
  
  # use top 500 genes (or length if less than 500)
  to_profile <- map(to_profile, ~.x[1:min(length(.x), 500)])
  
  go_res <- gost(to_profile, 
                 organism = "mmusculus",
                 significant = FALSE)
  
  go_res <- split(go_res$result, go_res$result$query)
  saveRDS(go_res, "go_results.rds")
}

go_res <- readRDS("go_results.rds")

iwalk(go_res,
     function(gres, id){
       x <- split(gres, gres$source)
       names(x) <- str_replace(names(x), ":", "_")
       write.xlsx(x, file.path(fig_dir, str_c("geneset_enrichment_results_", id, ".xlsx")))
     })
```


## Average expression per cluster


```{r}
combined_expression <- readRDS(file.path(formatted_data_dir, "avg_expression.rds"))


shared_genes <- map(combined_expression[mouse_studies[2:4]], ~.x$mouse_gene_name) %>% Reduce(intersect, .)

combined_expr <- imap(combined_expression[3:6], ~.x[!is.na(.x$mouse_gene_name), ] %>% 
      as.data.frame() %>% 
      filter(mouse_gene_name %in% shared_genes) %>% 
      dplyr::select(mouse_gene_name, contains(.y)) %>% 
      column_to_rownames("mouse_gene_name") %>% 
        .[shared_genes, , drop = FALSE]) %>% 
  unname() %>% 
  do.call(cbind, .)

combined_expr <- combined_expr[rowSums(is.na(combined_expr)) == 0, ]

# var_genes <- order(-apply(combined_expr, 1, var))[1:1000]
# combined_expr <- combined_expr[var_genes, ]

cres <- cor(combined_expr, method = "pearson")

hmap <- Heatmap(cres,
        col = viridis::viridis(256),
        name = "Correlation\n(Pearson)",
         row_names_gp = gpar(fontsize = 6),
        column_names_gp = gpar(fontsize = 6),
        heatmap_legend_param = list( labels_gp = gpar(fontsize = 6)))

hmap

pdf(file.path(fig_dir, "cell_type_correlation_mouse.pdf"),
    width = 10,
    height = 8)
draw(hmap, heatmap_legend_side = "left")
dev.off()
```

#### Specific marker heatmap

```{r}
to_plot <- c(
  "Sftpc",
  "Sftpd",
#  "C5",
  "Abca3",
  "Krt8",
  "Cldn4",
  "Sox4",
  "Itgb6",
  "Hopx",
  "Ager",
  "Akap5",
  "Pdpn")

zscore <- t(scale(t(combined_expr[to_plot, ])))

hmap <- Heatmap(zscore,
        col = viridis::viridis(256),
        name = "Z-score",
         row_names_gp = gpar(fontsize = 8),
        column_names_gp = gpar(fontsize = 8),
        row_names_side = "left",
        heatmap_legend_param = list( labels_gp = gpar(fontsize = 8)))

hmap


pdf(file.path(fig_dir, "cell_type_markers_heatmap_mouse.pdf"),
    width = 8,
    height = 8)
draw(hmap, heatmap_legend_side = "left")
dev.off()
```


### PCA plot

```{r}
library(ggrepel)
var_genes <- order(-apply(combined_expr, 1, var))[1:500]
zscores <- scale(t(combined_expr[var_genes, ]))
pc <- prcomp(zscores)

cell_types <- c("Joshi_et_al_european-r-j::AT2_2" = "Injured AT2",
  "Joshi_et_al_european-r-j::AT2_1" = "Injured AT2",
  "Joshi_et_al_european-r-j::AT2_3" = "CCA",
  "Joshi_et_al_european-r-j::AT2_4" = "Injured AT2",
  "Joshi_et_al_european-r-j::AT1" = "Naive AT1",
  "Riemondy_et_al_JCI-insight::Naive AEC1" = "Naive AT1",
  "Riemondy_et_al_JCI-insight::Naive AEC2" = "Naive AT2",
  "Riemondy_et_al_JCI-insight::Injured AEC2: Early Transdifferentiating" = "CCA",
  "Strunz_et_al_bioRxiv::Krt8_c4-epi-timecourse" = "CCA",
  "Strunz_et_al_bioRxiv::AT1-epi-timecourse" = "Naive AT1",
  "Strunz_et_al_bioRxiv::AT2-epi-timecourse" = "Naive AT2",
  "Strunz_et_al_bioRxiv::Krt8_c9-epi-timecourse" = "CCA",
  "Strunz_et_al_bioRxiv::AT2-whole-lung" = "Naive AT2",
  "Strunz_et_al_bioRxiv::Krt8-whole-lung" = "CCA",
  "Strunz_et_al_bioRxiv::AT1-whole-lung" = "Naive AT1",
  "Strunz_et_al_bioRxiv::AT2-injured-whole-lung" = "Injured AT2",
  "Wu_et_al_cell::population_I_d21" = "Injured AT2",
  "Wu_et_al_cell::population_II_d21" = "Injured AT2",
  "Wu_et_al_cell::AT1" = "Naive AT1",
  "Wu_et_al_cell::subpopulation I_d7" = "Injured AT2",
  "Wu_et_al_cell::subpopulation II_d7" = "Injured AT2")

cell_type_metadata <- tibble(
  study = names(combined_expr)) %>% 
  mutate(cell_types = cell_types[study])


pc$x %>% 
  as.data.frame() %>% 
  rownames_to_column("study") %>% 
  left_join(cell_type_metadata) %>% 
  ggplot(aes(PC1, PC2)) +
  geom_point(aes(color = cell_types)) 
```

## Try to co-embed all mouse data

```{r, harmonize, cache.lazy=FALSE}


so_fn <-  file.path(obj_dir, "harmonized.rds")

to_integrate <- c("Kobayashi_et_al_ncb",
                  "Riemondy_et_al_JCI-insight",
                  "Strunz_et_al_nc")

study_specific_cell_types <- list(
  "Kobayashi_et_al_ncb" = c(
    "AEC1",
    "AEC2",
    "PATS-2",
    "PATS-1",
    "AEC2-proliferating"
  ),
  "Riemondy_et_al_JCI-insight" =   c(
    "Naive AEC1",
    "Naive AEC2",
    "Injured AEC2: Early Transdifferentiating",
    "Injured AEC2: Late Transdifferentiating",
    "Injured AEC2: Proliferating",
    "Other Injured AEC2"
  ),
  "Strunz_et_al_nc" =  c(
    "Krt8 ADI",
    "Activated AT2 cells",
    "AT1 cells",
    "AT2 cells",
    "activated AT2",
    "proliferation"
  )
)

# study_specific_metadata <- list(
#   
# )
if(!file.exists(so_fn)){
  sos <- map(to_integrate, 
             ~readRDS(file.path(data_dir, .x, "so.rds")))
  names(sos) <- to_integrate
  sos <- pmap(list(sos, 
                   names(sos),
                   c("cell_type",
                     "pretty_cell_labels", 
                     "cell_type"),
                   study_specific_cell_types),
              function(so, id, cell_type_col, cell_types) {
                so@meta.data$study <- id
                so@meta.data$cell_type <- so@meta.data[[cell_type_col]]
                # so@meta.data <- so@meta.data[, 
                #                              c("orig.ident", "nCount_RNA", "nFeature_RNA", "study", "cell_type")]
                to_keep <- so$cell_type %in% cell_types
                so <- so[, to_keep]
                so})
  
  so <- merge(sos[[1]], y = sos)
  
  library(harmony)
  seed_value <- 20200806
  so <- FindVariableFeatures(so, nfeatures = 3000) %>% 
    ScaleData(.) %>% 
    RunPCA(., seed.use = seed_value) %>% 
    RunUMAP(., dims = 1:30, seed.use = seed_value)
  
  res_values <- seq(0.1, 1.5, by = 0.2)
  set.seed(seed_value)
  so <- RunHarmony(so, dims.use = 1:30, group.by.vars = "study")
  so <- RunUMAP(so, 
                dims = 1:30, 
                reduction = "harmony", 
                reduction.name = "harmony_umap", 
                seed.use = seed_value)
  
  so <- FindNeighbors(so, 
                      dims = 1:30, 
                      k.param = 30L, 
                      reduction = "harmony") %>% 
    FindClusters(resolution = res_values, 
                 random.seed = seed_value)
  
  # fix labels
  
  so$pretty_cell_labels <- case_when(
    so$study == "Riemondy_et_al_JCI-insight" ~ str_c("LPS:", so$cell_type),
    so$study == "Kobayashi_et_al_ncb" ~ str_c("Organoid:", so$cell_type),
    TRUE ~ so$pretty_cell_labels
  )
  saveRDS(so, file.path(obj_dir, "harmonized.rds"))
  rm(sos)
}
```

```{r, cache.lazy=FALSE}
so <- readRDS(file.path(obj_dir, "harmonized.rds"))
```


## Integrated UMAPs Krt8/18 markers {.tabset}

### Cell type not split
```{r, cache.lazy = FALSE}
to_plot <- c("pretty_cell_labels", "study")
umap_dir <- file.path(fig_dir, "umaps")
dir.create(umap_dir, showWarnings = FALSE)
pwalk(list(to_plot, 
      c(3.33, 1.75),
      c(2, 1)),
    function(x, y, z){
      p <- plot_harmony(so, x, sorted = "random", legend_title = "") +
        guides(color = guide_legend(ncol = z, override.aes = list(size = 4)))
      print(p)
      save_plot(file.path(umap_dir, str_c(x, "_not_split.pdf")),
                p,
                base_asp = y)
    })
```

```{r, results="asis", cache.lazy = FALSE}

iwalk(to_plot, 
    function(x, y){
      cat("\n### ", x, "\n")
      p <- plot_harmony(so, x, group = "study", sorted = "random", legend_title = "") +
        guides(color = guide_legend(ncol = c(2, 1)[y], override.aes = list(size = 4)))
      print(p)
      save_plot(file.path(umap_dir, str_c(x, "_split.pdf")),
                p,
                base_asp = c(5, 3)[y])
      cat('  \n')
    })
```

```{r, results="asis", cache.lazy = FALSE}
to_plot <- c("Hopx")

umap_dir <- file.path(fig_dir, "umaps")
dir.create(umap_dir, showWarnings = FALSE)
walk(to_plot, 
    function(x){
      cat("\n### ", x, "\n")
      p <- plot_harmony(so, x, group = "study", sorted = "random")
      print(p)
      save_plot(file.path(umap_dir, str_c(x, "_split.pdf")),
                p,
                base_asp = 1.5)
      cat('  \n')
    })
```

```{r, results="asis", cache.lazy = FALSE}
to_plot <- c("Hopx",
             "S100a6",
             "Sfn",
             "Tmsb10",
             "Anxa1",
             "Cldn4",
             "Clu",
             "AW112010",
             "Krt8",
             "Krt19",
             "Krt18",
             "Krt7",
             "Tnip3",
             "Tpm2",
             "Lgals1",
             "Sprr1a",
             "Anxa2",
             "Epcam",
             "Lgals3",
             "Hspb1",
             "2200002D01Rik",
             "Itgb1",
             "Sox4")

umap_dir <- file.path(fig_dir, "umaps")
dir.create(umap_dir, showWarnings = FALSE)
walk(to_plot, 
    function(x){
      cat("\n### ", x, "\n")
      p <- plot_harmony(so, x, group = "study", sorted = "random")
      print(p)
      save_plot(file.path(umap_dir, str_c(x, "_split.pdf")),
                p,
                base_asp = 3)
      cat('  \n')
    })
```
## Markers

```{r, cache.lazy = FALSE, fig.width = 10, fig.height= 10}
clustering_params <- str_subset(colnames(so@meta.data), "RNA_snn")
p <- plot_harmony(so, clustering_params) %>% 
  plot_grid(plotlist = .) 
p
save_plot(file.path(umap_dir, "clustering_parameters.pdf"),
          p,
          nrow = 3,
          ncol = 3,
          base_asp = 1.66)
```

```{r}
so$refined_clusters <- so$RNA_snn_res.1.3
so$coarse_clusters <- so$RNA_snn_res.0.5
saveRDS(so, file.path(obj_dir, "harmonized.rds"))

library(presto)
mkrs <- wilcoxauc(so, "refined_clusters")

top_mkrs <- mkrs %>% 
  filter(logFC > 0,
         padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE)

write_tsv(top_mkrs, file.path(fig_dir, "meta-analysis-refined-cluster-markers.tsv"))

mkrs <- wilcoxauc(so, "coarse_clusters")

top_mkrs <- mkrs %>% 
  filter(logFC > 0,
         padj < 0.05) %>% 
  group_by(group) %>% 
  arrange(padj, desc(logFC), .by_group = TRUE)

write_tsv(top_mkrs, file.path(fig_dir, "meta-analysis-coarse-cluster-markers.tsv"))
```

## Cell browser

```{r, eval = FALSE}
cb_outdir <- "cellbrowser"
dir.create(cb_outdir, showWarnings = FALSE)
so$study <- str_replace_all(so$study, "_", "-")
cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `cell-type` = "pretty_cell_labels",
  `study` = "study",
  `refined-clusters` = "refined_clusters",
  `coarse-clusters` = "coarse_clusters",
  `bleo-sample-info` = "grouping"
  )

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 project = "lung-meta-analysis",
                 outdir = cb_outdir,
                 marker_file = file.path(fig_dir, 
                                         "meta-analysis-refined-cluster-markers.tsv"),
                 ident = "refined-clusters",
                 embeddings = c("umap", "harmony_umap"),
                 skip_expr_matrix = TRUE,
                 config = list(priority = 1,  
                               radius = 2),
                 description = list(    
                   title = "Lung Injury meta-analysis",
                   description = "Includes unaligned and aligned (harmony_umap) projections"), 
                 cellbrowser_dir = "/miniconda3/envs/py37/bin"
)
```

```{r}
studies <- unique(so$study)
walk(studies, 
     ~{
       tmp <- subset(so, subset = study == .x)
       make_cellbrowser(tmp, 
                        column_list = cols_to_keep,
                        project = str_c("lung-meta-analysis-", .x) ,
                        outdir = cb_outdir,
                        marker_file = file.path(fig_dir, 
                                                "meta-analysis-refined-cluster-markers.tsv"),
                        ident = "refined-clusters",
                        embeddings = c("umap", "harmony_umap"),
                        skip_expr_matrix = FALSE,
                        config = list(priority = 1,  
                                      radius = 2),
                        description = list(    
                          title = .x,
                          description = "Includes unaligned and aligned (harmony_umap) projections"), 
                        cellbrowser_dir = "/miniconda3/envs/py37/bin"
       )})

```

```{r}
datasets <- file.path(cb_outdir,
                      c("lung-meta-analysis", str_c("lung-meta-analysis-", studies)),
                      "cellbrowser.conf")

build_cellbrowser(datasets, outdir = file.path(cb_outdir, "lung-meta"))
```


## Human


### Average expression per cluster


```{r}
combined_expression <- readRDS(file.path(formatted_data_dir, "avg_expression.rds"))

shared_genes <- map(combined_expression[1:2], 
                    ~.x$human_gene_name) %>% 
  Reduce(intersect, .)

combined_expr <- imap(combined_expression[1:2], 
                      ~.x[!is.na(.x$human_gene_name), ] %>% 
      as.data.frame() %>% 
      filter(human_gene_name %in% shared_genes) %>% 
      dplyr::select(human_gene_name, contains(.y)) %>% 
      column_to_rownames("human_gene_name") %>% 
        .[shared_genes, , drop = FALSE]) %>% 
  unname() %>% 
  do.call(cbind, .)

combined_expr <- combined_expr[rowSums(is.na(combined_expr)) == 0, ]

# var_genes <- order(-apply(combined_expr, 1, var))[1:1000]
# combined_expr <- combined_expr[var_genes, ]

cres <- cor(combined_expr, method = "pearson")

hmap <- Heatmap(cres,
        col = viridis::viridis(256),
        name = "Correlation\n(Pearson)",
         row_names_gp = gpar(fontsize = 6),
        column_names_gp = gpar(fontsize = 6),
        heatmap_legend_param = list( labels_gp = gpar(fontsize = 6)))

hmap

pdf(file.path(fig_dir, "cell_type_correlation_human.pdf"),
    width = 10,
    height = 8)
draw(hmap, heatmap_legend_side = "left")
dev.off()
```

### Specific marker heatmap

```{r}
to_plot <- c(
  "SFTPC",
  "SFTPD",
#  "C5",
  "ABCA3",
  "KRT8",
  "CLDN4",
  "SOX4",
  "ITGB6",
  "HOPX",
  "AGER",
  "AKAP5",
  "PDPN")

zscore <- t(scale(t(combined_expr[to_plot, ])))

hmap <- Heatmap(zscore,
        col = viridis::viridis(256),
        name = "Z-score",
         row_names_gp = gpar(fontsize = 8),
        column_names_gp = gpar(fontsize = 8),
        row_names_side = "left",
        heatmap_legend_param = list( labels_gp = gpar(fontsize = 8)))

hmap


pdf(file.path(fig_dir, "cell_type_markers_heatmap_human.pdf"),
    width = 8,
    height = 8)
draw(hmap, heatmap_legend_side = "left")
dev.off()
```


## Average expression per cluster for both


```{r}
combined_expression <- readRDS(file.path(formatted_data_dir, "avg_expression.rds"))

shared_genes <- map(combined_expression, 
                    ~.x$ortho_id) %>% 
  Reduce(intersect, .)

combined_expr <- imap(combined_expression, 
                      ~.x[!is.na(.x$ortho_id), ] %>% 
      as.data.frame() %>% 
      filter(ortho_id %in% shared_genes) %>% 
      dplyr::select(ortho_id, contains(.y)) %>% 
      column_to_rownames("ortho_id") %>% 
        .[shared_genes, , drop = FALSE]) %>% 
  unname() %>% 
  do.call(cbind, .)

combined_expr <- combined_expr[rowSums(is.na(combined_expr)) == 0, ]

# var_genes <- order(-apply(combined_expr, 1, var))[1:1000]
# combined_expr <- combined_expr[var_genes, ]

cres <- cor(combined_expr, method = "pearson")

hmap <- Heatmap(cres,
        col = viridis::viridis(256),
        name = "Correlation\n(Pearson)",
         row_names_gp = gpar(fontsize = 6),
        column_names_gp = gpar(fontsize = 6),
        heatmap_legend_param = list( labels_gp = gpar(fontsize = 6)))

hmap

pdf(file.path(fig_dir, "cell_type_correlation_human_mouse.pdf"),
    width = 10,
    height = 8)
draw(hmap, heatmap_legend_side = "left")
dev.off()
```



### Specific marker heatmap

```{r}
to_plot <- c(
  "SFTPC",
  "SFTPD",
#  "C5",
  "ABCA3",
  "KRT8",
  "CLDN4",
  "SOX4",
  "ITGB6",
  "HOPX",
  "AGER",
  "AKAP5",
  "PDPN")

to_plot <- c("SFTPC:Sftpc",
  "SFTPD:Sftpd",
  "ABCA3:Abca3",
  "KRT8:Krt8",
  "CLDN4:Cldn4",
  "SOX4:Sox4",
  "ITGB6:Itgb6",
  "HOPX:Hopx",
  "AGER:Ager",
  "AKAP5:Akap5",
  "PDPN:Pdpn")

zscore <- t(scale(t(combined_expr[to_plot, ])))

hmap <- Heatmap(zscore,
        col = viridis::viridis(256),
        name = "Z-score",
         row_names_gp = gpar(fontsize = 8),
        column_names_gp = gpar(fontsize = 8),
        row_names_side = "left",
        heatmap_legend_param = list( labels_gp = gpar(fontsize = 8)))

hmap


pdf(file.path(fig_dir, "cell_type_markers_heatmap_human_mouse.pdf"),
    width = 8,
    height = 8)
draw(hmap, heatmap_legend_side = "left")
dev.off()
```


### Marker overlap

```{r}
library(eulerr)

human_markers <- map2(combined_tables[human_studies], 
                     c("gene", "X1", "id"),
                     ~.x[[.y]])
fit <- euler(human_markers)
pdf(file.path(fig_dir, "marker_overlap_human_studies.pdf"))
plot(fit, 
     labels = TRUE,
     quantities = TRUE,
     fills = list(fill = palette_OkabeIto[1:6],
                  alpha = 0.5))
dev.off()

markers_shared_human <- Reduce(intersect, human_markers)

#get mouse ortholog for human gene
markers_shared_human <- left_join(tibble(human_gene_name = markers_shared_human),
          orthologs,
          by = c("human_gene_name"))

all_human_markers <- unlist(human_markers, use.names = FALSE) %>% unique()
                      
mrkr_lgl_human <- imap(human_markers,
    function(x, id) {
      df <- data.frame(tmp = all_human_markers %in% x, 
                       row.names = all_human_markers)
      colnames(df) <- id
      df}) %>% 
  do.call(cbind, .) %>% 
  rownames_to_column("gene")

markers_shared_mouse <- Reduce(intersect, combined_markers[mouse_studies])
# markers_shared_mouse_orthos <- left_join(tibble(mouse_gene_name = markers_shared_mouse),
#           orthologs,
#           by = c("mouse_gene_name"))

all_mouse_markers <- unlist(combined_markers[mouse_studies], use.names = FALSE) %>% unique()
                      
mrkr_lgl <- imap(combined_markers[mouse_studies],
    function(x, id) {
      df <- data.frame(tmp = all_mouse_markers %in% x, 
                       row.names = all_mouse_markers)
      colnames(df) <- id
      df}) %>% 
  do.call(cbind, .)

markers_3_of_4 <- mrkr_lgl[rowSums(mrkr_lgl) == 3, ] %>% 
  rownames_to_column("gene")

openxlsx::write.xlsx(list(
  shared_mouse_markers = markers_shared_mouse,
  mouse_shared_3_of_4 = markers_3_of_4,
  shared_human_markers = markers_shared_human,
  human_all_data = mrkr_lgl_human),
     file.path(fig_dir, "shared_genes.xlsx"))

```
