---
title: "Aggregte gene lists"
author: "Kent Riemondy RBI"
date: "2/24/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
```

```{r}
library(tidyverse)
library(Seurat)
library(here)
library(readxl)
library(scbp)
data_dir <- here("data")
```

## Get ortholog table

```{r}
ortholog_table <- file.path(data_dir, "orthologs", "mouse_human_orthologs.tsv")

if(!file.exists(ortholog_table)){
  dir.create(file.path(data_dir, "orthologs"), showWarnings = FALSE)
  
  library(biomaRt)
  
  ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  
  orthologs <- getBM(mart = ensembl, 
      attributes = c("external_gene_name",
                     "mmusculus_homolog_associated_gene_name"))
  
  orthologs <- filter(orthologs,!is.na(mmusculus_homolog_associated_gene_name)) %>%
    group_by(external_gene_name) %>%
    mutate(n_orthos_h = n()) %>%
    group_by(mmusculus_homolog_associated_gene_name) %>%
    mutate(n_orthos_m = n()) %>%
    filter(n_orthos_h == 1, n_orthos_m == 1) %>%
    dplyr::select(-starts_with("n_orthos")) %>%
    ungroup()
  
  orthologs <- mutate(orthologs, 
                      ortho_id = str_c(external_gene_name, ":", mmusculus_homolog_associated_gene_name)) %>% 
    dplyr::rename(human_gene_name = external_gene_name,
                  mouse_gene_name = mmusculus_homolog_associated_gene_name)

  write_tsv(orthologs, ortholog_table)
}

orthologs <- read_tsv(
  ortholog_table,
  col_types = cols(
    human_gene_name = col_character(),
    mouse_gene_name = col_character(),
    ortho_id = col_character()
  )
)

```

## Gene lists

```{r}
studies <- c(
  "Adams_et_al_bioRxiv" = "Data_S1.xlsx",
  "Habermann_et_al_bioRxiv" = "KRT17_vs_allepi.csv",
  "Reyfman_et_al_ajrccm" = "3.tsv.gz",
  "Joshi_et_al_european-r-j" = "AT2_clean.markers.csv",
  "Riemondy_et_al_JCI-insight" = "2.14.2019_Suppl_Table_5.xlsx",
  "Strunz_et_al_bioRxiv" = "shiller_krt8_markers.xlsx",
  "Wu_et_al_cell" = "Nan Tang Table S2.xlsx"
  )

studies <- imap(studies, ~file.path(data_dir, .y, .x))

combined_markers <- list()
combined_tables <- list()
combined_expression <- list()
```

## Adams_et_al_bioRxiv

```{r}
mkrs <- studies[[1]]
study_id <- names(studies[1])

tbl <- read_excel(mkrs) %>% 
  filter(cellType == "Aberrant Basaloid") %>% 
  arrange(wilcox.fdr) %>% 
  filter(wilcox.fdr < 0.05,
         log1p_FC > 0)

mouse_markers <- left_join(tbl, orthologs, by = c("gene" = "human_gene_name")) %>% 
  pull(mouse_gene_name) %>% 
  .[!is.na(.)]

combined_markers[[study_id]] <- mouse_markers
combined_tables[[study_id]] <- tbl
```

## Habermann_et_al_bioRxiv

```{r}
mkrs <- studies[[2]]
study_id <- names(studies[2])

tbl <- read_csv(mkrs) %>% 
  arrange(p_val_adj) %>% 
  filter(p_val_adj < 0.05,
         avg_logFC > 0)

mouse_markers <- left_join(tbl, orthologs, by = c("X1" = "human_gene_name")) %>% 
  pull(mouse_gene_name) %>% 
  .[!is.na(.)]

combined_markers[[study_id]] <- mouse_markers
combined_tables[[study_id]] <- tbl

so <- readRDS(file.path(data_dir, study_id, "GSE135893_ILD_annotated_diet.rds"))
so_sub <- subset(so, celltype %in% c("KRT5-/KRT17+", "AT1", "AT2", "Transitional AT2")) 
so_sub <- NormalizeData(so_sub, assay = "RNA")
avg_expr <- AverageExpression(so_sub)$RNA
colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "human_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "human_gene_name")

combined_expression[[study_id]] <- avg_expr

rm(so, so_sub); gc()
```

## Reyfmann_et_al_ajrccm

```{r}
mkrs <- studies[[3]]
study_id <- names(studies[3])

tbl <- read_tsv(mkrs) %>% 
  arrange(`p_val|float`, desc(`avg_diff|float`)) %>% 
  filter(`p_val|float` < 0.05,
         `avg_diff|float` > 0)

mouse_markers <- left_join(tbl, orthologs, by = c("symbol" = "human_gene_name")) %>% 
  pull(mouse_gene_name) %>% 
  .[!is.na(.)]

combined_markers[[study_id]] <- mouse_markers
combined_tables[[study_id]] <- tbl

so_fn <- file.path(data_dir, "Reyfman_et_al_ajrccm", "so.rds")

if(!file.exists(so_fn)){
  mat <- read_tsv(file.path(data_dir, "Reyfman_et_al_ajrccm", "exprMatrix.tsv.gz"))
  mat <- column_to_rownames(mat, "gene")
  mat <- as.data.frame(mat) %>% as.matrix()
  mat <- as(mat, "sparseMatrix")
  
  # mat_url <- "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3489182&format=file&file=GSM3489182%5FDonor%5F01%5Fraw%5Fgene%5Fbc%5Fmatrices%5Fh5%2Eh5"
  # mat_fn <- file.path(data_dir, "Reyfman_et_al_ajrccm", "matrix.h5")
  # download.file(mat_url, mat_fn)
  # so <- Read10X_h5(mat_fn)
  
  mdata <- read_tsv(file.path(data_dir, "Reyfman_et_al_ajrccm", "meta.tsv")) %>% 
    column_to_rownames("Cell")
  so <- CreateSeuratObject(mat, meta.data = mdata)
  saveRDS(so, so_fn)
}
so <- readRDS(so_fn)
Idents(so) <- "Cluster"
avg_expr <- AverageExpression(so)$RNA
colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "human_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "human_gene_name")

combined_expression[[study_id]] <- avg_expr
rm(so); gc()
```

## Joshi_et_al_european-r-j

```{r}
mkrs <- studies[[4]]
study_id <- names(studies[4])

tbl <- read_csv(mkrs) %>% 
  filter(cluster == "AT2_3") %>% 
  arrange(p_val_adj) %>% 
  filter(p_val_adj < 0.05,
         avg_logFC > 0)

combined_markers[[study_id]] <-tbl$X1
combined_tables[[study_id]] <- tbl

so_fn <- file.path(data_dir, study_id, "so.rds")
if(!file.exists(so_fn)){
  load(file.path(data_dir, study_id, "AT2_clean.Robj"))
  saveRDS(AT2_clean, so_fn)
}

so <- readRDS(so_fn)

avg_expr <-  AverageExpression(so)$RNA
#avg_expr <- avg_expr[, "AT2_3", drop = FALSE]
colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "mouse_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "mouse_gene_name")

combined_expression[[study_id]] <- avg_expr

rm(so); gc()
```

## Riemondy_et_al_JCI-insight

```{r}
mkrs <- studies[[5]]
study_id <- names(studies[5])

tbl <- read_excel(mkrs, sheet = 2,
                  col_names = c(
                    "Gene",
                    "avg_logFC",
                    "pct1",
                    "pct2",
                    "pval",
                    "adj_p_val"
                  ), skip = 1) %>% 
  arrange(adj_p_val) %>% 
  filter(adj_p_val < 0.05)

combined_markers[[study_id]] <-tbl$Gene
combined_tables[[study_id]] <- tbl

so_fn <- file.path(data_dir, study_id, "so.rds")
if(!file.exists(so_fn)){
  ## update old seurat file
  so <- readRDS(file.path(data_dir, study_id, "s_obj_final.rds"))
  so <- UpdateSeuratObject(so)
  mdata <- so@meta.data
  new_mdata <- tibble::rownames_to_column(mdata, "cell") %>% 
    dplyr::mutate(sample_type = ifelse(str_detect(new_expt_id, 
                                                  "Non-ATII epithelial"),
                                       "Naive Non-AEC2 Epithelial",
                                       ifelse(str_detect(new_expt_id,
                                                         "ATII expt"),
                                              "Naive AEC2",
                                              "Injured AEC2")),
                  experiment = ifelse(expt == "expt1", 
                                      "Experiment 1",
                                      "Experiment 2")) %>% 
    dplyr::select(cell, sample_type, experiment) %>% 
    tibble::column_to_rownames("cell")
  
  so <- AddMetaData(so, new_mdata)
  
  so <- AddMetaData(so, 
                    data.frame(row.names = rownames(so@meta.data),
                               technical_replicate = str_match(so@meta.data$sample_names, "([0-9])expt")[, 2]))
  
  so <- AddMetaData(so,
                    data.frame(row.names = rownames(so@meta.data),
                               experiment_and_tech_rep = str_c(so@meta.data$experiment,
                                                               ": Technical Replicate ",
                                                               so@meta.data$technical_replicate)))
  
  cell_id_relabel <- c("Endothelial" = "Endothelial/Fibroblast",
                       "Basal" = "Basal",
                       "Club" = "Club",
                       "Ciliary" = "Ciliated",
                       "Naive Type I" = "Naive AEC1",
                       "Macrophages" = "Macrophage",
                       "Injured Type II" = "Other Injured AEC2",
                       "Proliferating Type II" = "Injured AEC2: Proliferating",
                       "Cell Cycle Arrest Type II" = "Injured AEC2: Early Transdifferentiating",
                       "Transdifferentiating Type II" = "Injured AEC2: Late Transdifferentiating",
                       "Naive Type II" = "Naive AEC2")
  
  
  m_df <- data.frame(row.names = rownames(so@meta.data),
                     pretty_cell_labels = cell_id_relabel[so@meta.data$labeled_clusters])
  
  m_df$pretty_cell_labels <- factor(m_df$pretty_cell_labels,
                                    levels = cell_id_relabel)
  
  so <- AddMetaData(so, m_df)
  saveRDS(so, file.path(data_dir, study_id, "so.rds"))
}

so <- readRDS(so_fn)
Idents(so) <- "pretty_cell_labels"
avg_expr <-  AverageExpression(so)$RNA
avg_expr <- avg_expr[, c("Naive AEC1", "Naive AEC2", "Injured AEC2: Early Transdifferentiating"), drop = FALSE]
colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "mouse_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "mouse_gene_name")

combined_expression[[study_id]] <- avg_expr

rm(so); gc()
```

## Strunz_et_al_bioRxiv

```{r}
mkrs <- studies[[6]]
study_id <- names(studies[6])

tbl <- read_excel(mkrs) %>% 
  arrange(p_val_adj) %>% 
  filter(p_val_adj < 0.05,
         avg_logFC > 0)

combined_markers[[study_id]] <- tbl$gene
combined_tables[[study_id]] <- tbl

#figure 4 data, need to find updated meta.data labels or annotate ourselves
# load(file.path(data_dir, study_id, "data",
#                "EpiHiRes_seurat.RData"))

so_fn <- file.path(data_dir, study_id, "so.rds")
if(!file.exists(so_fn)){
  # figure 2, matches publication
  load(file.path(data_dir, study_id, "data",
                 "Whole_lung_seurat_alveolar_epithelium.RData"))
  
  so <- UpdateSeuratObject(seu)
  saveRDS(so, file.path(data_dir, study_id, "so.rds"))
}
so <- readRDS(file.path(data_dir, study_id, "so.rds"))
Idents(so) <- "cell_type"
avg_expr <-  AverageExpression(so)$RNA

colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr))
avg_expr <- rownames_to_column(avg_expr, "mouse_gene_name")
avg_expr <- left_join(avg_expr, orthologs, by = "mouse_gene_name")

combined_expression[[study_id]] <- avg_expr

rm(seu, so); gc()
```

## Wu_et_al_cell

```{r}
mkrs <- studies[[7]]
study_id <- names(studies[7])

tbl <- read_excel(mkrs, skip = 1) %>% 
  filter(`fold change (subpopulation I AT2 cells/subpopulation II AT2 cells)` > 0)

combined_markers[[study_id]] <- tbl$`Gene name`
combined_tables[[study_id]] <- tbl

avg_expr_d7 <- read_excel(file.path(data_dir, names(studies)[7], "C7.average.exp.xlsx"))
avg_expr_d7 <- dplyr::rename(avg_expr_d7, 
                          mouse_gene_name = "...1")

colnames(avg_expr_d7)[2:3] <- paste0(study_id, "::", colnames(avg_expr_d7)[2:3], "_d7")

so_fn <- file.path(data_dir, study_id, "so.rds")
if(!file.exists(so_fn)){
  # also process d21 data
  mat <- read_csv(file.path(data_dir,  names(studies)[7], "matrix.tsv.gz"))
  mat <- column_to_rownames(mat, "X1") 
  mat <- as.matrix(mat)
  mat <- as(mat, "sparseMatrix")
  
  mdata <- read_csv(file.path(data_dir,  names(studies)[7], "meta_info_C0_C7_C21.csv")) %>% 
    dplyr::select(cell = X1, sample_name, cell_type) %>% 
    filter(cell %in% colnames(mat)) %>% 
    as.data.frame() %>% 
    column_to_rownames("cell") 
  
  mat <- mat[, rownames(mdata)]
  so <- CreateSeuratObject(mat, meta.data = mdata) %>% 
    NormalizeData()
  saveRDS(so, so_fn)
}

so <- readRDS(so_fn)
Idents(so) <- "cell_type"
avg_expr <- AverageExpression(so)$RNA
#avg_expr <- avg_expr[, "population II", drop = FALSE]
colnames(avg_expr) <- paste0(study_id, "::", colnames(avg_expr), "_d21")
avg_expr <- rownames_to_column(avg_expr, "mouse_gene_name")
avg_expr_d21 <- left_join(avg_expr, orthologs, by = "mouse_gene_name")
avg_expr <- left_join(avg_expr_d21, avg_expr_d7, by = "mouse_gene_name")
combined_expression[[study_id]] <- avg_expr
```


## Save 

```{r}
formatted_data_dir <- "processed_data"
dir.create(formatted_data_dir, showWarnings = FALSE)
saveRDS(combined_markers, file.path(formatted_data_dir, "gene_lists.rds"))
saveRDS(combined_tables, file.path(formatted_data_dir, "gene_tables.rds"))
saveRDS(combined_expression, file.path(formatted_data_dir, "avg_expression.rds"))

```

```{r}
fig_dir <- "figs"
dir.create(fig_dir, showWarnings = FALSE)
combined_markers <- readRDS(file.path('processed_data', "gene_lists.rds"))

library(eulerr)

fit <- venn(combined_markers[3:6])
pdf(file.path(fig_dir, "marker_overlap_mouse_studies.pdf"))
plot(fit, 
     labels = TRUE,
     quantities = TRUE,
     fills = list(fill = palette_OkabeIto[1:4],
                  alpha = 0.5))
dev.off()


fit <- euler(combined_markers[1:2])
pdf(file.path(fig_dir, "marker_overlap_human_studies.pdf"))
plot(fit, 
     labels = TRUE,
     quantities = TRUE,
     fills = list(fill = palette_OkabeIto[1:6],
                  alpha = 0.5))
dev.off()

```

```{r}
library(UpSetR)

upset(fromList(combined_markers[3:6]))
pdf(file.path(fig_dir, "marker_overlap_mouse_studies_upset_plot.pdf"))
  upset(fromList(combined_markers[3:6]))
dev.off()

upset(fromList(combined_markers), nsets = 6, nintersect = 50)
pdf(file.path(fig_dir, "marker_overlap_all_studies_upset_plot.pdf"))
  upset(fromList(combined_markers))
dev.off()
```


## Average expression per cluster
```{r, eval = FALSE}
combined_expression <- readRDS(file.path(formatted_data_dir, "avg_expression.rds"))


shared_genes <- map(combined_expression[3:6], ~.x$mouse_gene_name) %>% Reduce(intersect, .)

combined_expr <- imap(combined_expression[3:6], ~.x[!is.na(.x$mouse_gene_name), ] %>% 
      as.data.frame() %>% 
      filter(mouse_gene_name %in% shared_genes) %>% 
      dplyr::select(mouse_gene_name, contains(.y)) %>% 
      column_to_rownames("mouse_gene_name") %>% 
        .[shared_genes, , drop = FALSE]) %>% 
  do.call(cbind, .)

combined_expr <- combined_expr[rowSums(is.na(combined_expr)) == 0, ]

cres <- cor(combined_expr, method = "pearson")
library(ComplexHeatmap)
Heatmap(cres)



shared_genes <- map(combined_expression[1:2], ~.x$human_gene_name) %>% Reduce(intersect, .)

combined_expr <- imap(combined_expression[1:2], ~.x[!is.na(.x$human_gene_name), ] %>% 
      as.data.frame() %>% 
      filter(human_gene_name %in% shared_genes) %>% 
      dplyr::select( human_gene_name, contains(.y)) %>% 
      column_to_rownames("human_gene_name") %>% 
        .[shared_genes, , drop = FALSE]) %>% 
  do.call(cbind, .)

combined_expr <- combined_expr[rowSums(is.na(combined_expr)) == 0, ]

cres <- cor(combined_expr, method = "pearson")
library(ComplexHeatmap)
Heatmap(cres)
```


## Try to co-embed all mouse data

```{r}

obj_dir <- "objects"
dir.create(obj_dir)
so_fn <-  file.path(obj_dir, "harmonized.rds")

if(!file.exists(so_fn)){
  mouse_studies <- names(studies[4:7])
  sos <- map(mouse_studies, 
             ~readRDS(file.path(data_dir, .x, "so.rds")))
  names(sos) <- mouse_studies
  sos <- imap(sos, function(so, id) {
    so@meta.data$study <- id
    so@meta.data <- so@meta.data[, c("orig.ident", "nCount_RNA", "nFeature_RNA", "study")]
    so})
  
  so <- merge(sos[[1]], y = sos)
  
  library(harmony)
  seed_value <- 20200331
  so <- FindVariableFeatures(so, nfeatures = 3000) %>% 
    ScaleData(.) %>% 
    RunPCA(., seed.use = seed_value) %>% 
    RunUMAP(., dims = 1:30, seed.use = seed_value)
  
  res_values <- seq(0.1, 1.5, by = 0.2)
  set.seed(seed_value)
  so <- RunHarmony(so, dims.use = 1:30, group.by.vars = "study", )
  so <- RunUMAP(so, dims = 1:30, reduction = "harmony", reduction.name = "harmony_umap", 
                seed.use = seed_value)
  
  so <- FindNeighbors(so, dims = 1:30) %>% 
    FindClusters(resolution = res_values, 
                 random.seed = seed_value)
  
  
  saveRDS(so, file.path(obj_dir, "harmonized.rds"))
}
```

```{r, cache.lazy=FALSE}
so <- readRDS(file.path(obj_dir, "harmonized.rds"))
```


## Integrated UMAPs Krt8/18 markers {.tabset}

```{r, cache.lazy = FALSE}
to_plot <- c("S100a6",
             "Sfn",
             "Tmsb10",
             "Anxa1",
             "Cldn4",
             "Clu",
             "AW112010",
             "Krt8",
             "Krt19",
             "Krt18",
             "Krt7",
             "Tnip3",
             "Tpm2",
             "Lgals1",
             "Sprr1a",
             "Anxa2",
             "Epcam",
             "Lgals3",
             "Hspb1",
             "2200002D01Rik")
```

```{r, results="asis", cache.lazy = FALSE}
walk(to_plot, 
    function(x){
      cat("\n### ", x, "\n")
      p <- plot_harmony(so, x, group = "study", sorted = "random")
      print(p)
      cat('  \n')
    })
```